### ISSUE TITLE
Why can't I get the gradient when I use in cascaded vmap?

### ISSUE THREAD

**Issue by Kyung-mi** [patrick-kidger/equinox#442]
Hello, I have a question about getting gradients with equinox.

In the function of **first** 'grad_and_loss' in the code below, when I tried to get the gradient but only got zero gradient.
For convenience, I used jax.vmap along batch size etc.

However, when I tried to get the gradient using **second** 'grad_and_loss', which contains function 'calc_sample' directly, I can get the nonzero gradient.
Isn't @eqx.filter_value_and_grad available when I used jax.vmap?? My code is simple, so I cannot find the reason why.

I'm looking forward to your response.
Thank you.

`def probability(s,x,t,y,sigma,b):
    # D : data size
    error = y - (x+(t-s)*b) # (D,)
    D = error.shape[0]
    error = jnp.atleast_2d(error)
    a = jnp.atleast_2d( sigma @ sigma.T ) # a = (D,D)
    p = (2*math.pi)**(-D/2) * jnp.abs(jnp.linalg.det(a))**(-1/2) * jnp.exp(-1/2*error.T * jnp.linalg.inv(a)*error)
    return p

model = NeuralODE(data_size, width_size, depth, mm_scaler, key=model_key)

@eqx.filter_jit
def diffusion_scaled(t, y, args):
    sigma = 0.15
    diff = sigma * ( mm_scaler.value_mat_inv * y) # change y into original scale
    diff = mm_scaler.value_mat * diff # scaled for derivative
    return diff

@eqx.filter_jit
def calc_sample(ti, yi, key):
    # print("calc sample ti : ", ti.shape)
    # print("calc sample yi : ", yi.shape)
    L = yi.shape[0]
    ti_N = jnp.zeros((L-1,N))
    for i in range(L-1):
        ti_N = ti_N.at[i,:].set(jnp.linspace( ti[i], ti[i+1], N ))
    ts, ys = jax.vmap(model.stochastic, in_axes=(0, 0, None))(ti_N, yi[:-1,:], key) # (L-1,N) (L-1,D)
    
    y_n1 = ys[:,-2,:].copy() # ys = (L-1,N,D) -> y_n1 = (L-1,1,D) at tau_(n-1)
    y_n = yi[1:,:].copy() # true value at tat[i,:].set(jnp.linspace( ti[i], ti[i+1], N ))

    # print("y_n1 : ", y_n1.shape)
    sigma = jax.vmap(diffusion_scaled)([],y_n1,[])
    b = jax.vmap(model.g1.mlp)(y_n1)
    
    prob_map = jax.vmap(probability, in_axes=(0,0,0,0,0,0))

    approx_p = prob_map(ts[:,-2], y_n1, ts[:,-1], y_n, sigma, b ) # probability (s,x,t,y,a,b) ; s < t

    return approx_p

@eqx.filter_jit
def calc_prob(ti, yi):
    # calculate sample mean probability for each batch
    # ti : (L,)
    # yi : (L,D)

    L = yi.shape[1]
    # sample_p = jnp.zeros((L,M))
    
    start = time.time()
    subkeys = jax.random.split(key, M) 
    # sample_p, penalty = calc_sample(ti, yi, key)
    sample_p = jax.vmap(calc_sample, in_axes=(None,None,0))(ti,yi,subkeys)
    
    sample_p = calc_sample(ti,yi,key)
    print("sample_p : ", sample_p.shape) # (M,L,D,1)
    end = time.time()
    
    print("Time : ", end-start)
    # sample_p_mean = jnp.nanmean(sample_p, axis=0).squeeze() # (M,L,D,1) -> (L,D)
    sample_p_mean = jnp.sum(sample_p)

    return sample_p_mean

# @eqx.filter_value_and_grad
# def grad_loss(model, ti, yi):
#     # yi : (batch_num, L=length * length_size, data size)
#     # ti : (L,)
#     # approximate the probability
    
#     # Sample M
#     p_pred = jax.vmap(calc_prob, in_axes=(None,0))(ti,yi) # along batch_num
#     return -jnp.sum( jnp.log(p_pred) ) # negative log likelihood

@eqx.filter_value_and_grad
def grad_loss(model, ti, yi):
    # yi : (batch_num, L=length * length_size, data size)
    # ti : (L,)
    
    L = yi.shape[0]
    ti_N = jnp.zeros((L-1,N))
    for i in range(L-1):
        ti_N = ti_N.at[i,:].set(jnp.linspace( ti[i], ti[i+1], N ))
    ts, ys = jax.vmap(model.stochastic, in_axes=(0, 0, None))(ti_N, yi[:-1,:], key) # (L-1,N) (L-1,D)
    
    y_n1 = ys[:,-2,:].copy() # ys = (L-1,N,D) -> y_n1 = (L-1,1,D) at tau_(n-1)
    y_n = yi[1:,:].copy() # true value at tat[i,:].set(jnp.linspace( ti[i], ti[i+1], N ))
    
    sigma = jax.vmap(diffusion_scaled)([],y_n1,[])
    b = jax.vmap(model.g1.mlp)(y_n1)
    
    prob_map = jax.vmap(probability, in_axes=(0,0,0,0,0,0))
    approx_p = prob_map(ts[:,-2], y_n1, ts[:,-1], y_n, sigma, b )# probability (s,x,t,y,a,b) ; s < t
 
    return -jnp.sum( jnp.log(approx_p) ) # negative log likelihood

@eqx.filter_jit
def make_step(ti, yi, model, opt_state):
    loss, grads = grad_loss(model, ti, yi[0,:,:])
    updates, opt_state = optim.update(grads, opt_state)
    model = eqx.apply_updates(model, updates)
    return loss, model, opt_state, grads

# Learning for different length size
for lr, steps, length in zip(lr_strategy, steps_strategy, length_strategy):

    optim = optax.adabelief(lr)
    opt_state = optim.init(eqx.filter(model, eqx.is_inexact_array))
    _ts = ts[: int(length_size * length)]
    _ys = ys[:, : int(length_size * length)]

    # _ys : (sample_num, length * length_size, data size)
    for step, (yi,) in zip(
        range(steps), dataloader((_ys,), batch_size, key=loader_key)
    ):
        # yi : (batch_num, length * length_size, data size)
        start = time.time()
        loss, model, opt_state, grads = make_step(_ts, yi, model, opt_state)
        
        end = time.time()
        if (step % print_every) == 0 or step == steps - 1:
            print(f"Step: {step}, Loss: {loss}, Computation time: {end - start}")
    break`

**Comment by patrick-kidger** [comment]
Hi there! This is quite a large code sample, and I'm afraid it doesn't run directly. Can you reproduce the issue in something smaller, please?

**Comment by Kyung-mi** [comment]
I changed my code as you can see above code for reproducing my problem.
Thank you for your quick response.

**Comment by patrick-kidger** [comment]
Hi! This is still a huge example. I'm afraid it would still take me a long time to dig through your code like this.

Please can you reduce this to a single code snippet, ideally <30 lines or so, demonstrating the issue?

**Comment by Kyung-mi** [comment]
``` python
import time
import math
import diffrax
import equinox as eqx  # https://github.com/patrick-kidger/equinox
import jax
import jax.nn as jnn
import jax.numpy as jnp
from jax import grad, jit, vmap
import matplotlib.pyplot as plt
import optax  # https://github.com/deepmind/optax
import jax.tree_util as jtu
import numpy as np
import pandas as pd

dataset_size=10
N=32
M=64 # sample number
batch_size=1 # batch data among total data (dataset_size)
width_size=64
depth=2
seed=5678

key = jax.random.PRNGKey(seed)
data_key, model_key, loader_key = jax.random.split(key, 3)

# data
length = 50
ts = jnp.linspace(0.0, 10.0, length)
ys = jnp.ones((10,length,1)) # batch, length, data size
_, length_size, data_size = ys.shape

class Func(eqx.Module):
    mlp: eqx.nn.MLP

    def __init__(self, data_size, width_size, depth, *, key, **kwargs):
        super().__init__(**kwargs)
        self.mlp = eqx.nn.MLP(
            in_size = data_size,
            out_size = data_size,
            width_size = width_size,
            depth = depth,
            activation = jnp.tanh,
            use_bias = True, # remove bias (False)
            use_final_bias = True,
            key = key,
        )

    def __call__(self, t, y, args):
        return self.mlp(y)
    
model = Func(data_size, width_size, depth, key=model_key) # MLP model

def probability(b):
    p = b ** 2
    return p

# L : Time length, D : Data dimension, batch_num : number of batch
@eqx.filter_jit
def calc_sample(ti, yi, key):
    # ti : (L,)
    # yi : (L,D)

    b = jax.vmap(model.mlp)(yi) # (L,D)
    approx_p = jax.vmap( probability )(b) # (L,D)

    return approx_p

@eqx.filter_jit
def calc_prob(ti, yi):
    # calculate sample mean probability for each batch
    # ti : (L,)
    # yi : (L,D)

    start = time.time()
    subkeys = jax.random.split(key, M) 
    sample_p = jax.vmap(calc_sample, in_axes=(None,None,0))(ti,yi,subkeys) # (M,L,D,1) 
    # sample_p = calc_sample(ti,yi,key) 
    end = time.time()
    
    print("Time : ", end-start)
    sample_p_mean = jnp.sum(sample_p)

    return sample_p_mean

@eqx.filter_value_and_grad
def grad_loss(model, ti, yi):
    # yi : (batch_num, L=length * length_size, data size)
    # ti : (L,)

    b = jax.vmap(jax.vmap(model.mlp))(yi) # (batch_num, L, D)
    approx_p = jax.vmap( jax.vmap(probability) )(b) # (batch_num, L, D) 
    return jnp.sum( approx_p )

@eqx.filter_jit
def make_step(ti, yi, model, opt_state):
    loss, grads = grad_loss(model, ti, yi)
    updates, opt_state = optim.update(grads, opt_state)
    model = eqx.apply_updates(model, updates)
    return loss, model, opt_state, grads

lr = 1e-3
steps = 100
optim = optax.adabelief(lr)
opt_state = optim.init(eqx.filter(model, eqx.is_inexact_array))

for step in range(steps):
    loss, model, opt_state, grads = make_step(ts, ys, model, opt_state)
    a,b,c = grads.mlp.layers
    print(a.weight[:3]) # check the gradients of weights in layers
```

If I try this one, then I got
![image](https://github.com/patrick-kidger/equinox/assets/54544508/f5ebbbca-bda5-4469-bfa1-5da61593433d)


However, I try this one (I just changed function 'grad_loss')

``` python
import time
import math
import diffrax
import equinox as eqx  # https://github.com/patrick-kidger/equinox
import jax
import jax.nn as jnn
import jax.numpy as jnp
from jax import grad, jit, vmap
import matplotlib.pyplot as plt
import optax  # https://github.com/deepmind/optax
import jax.tree_util as jtu
import numpy as np
import pandas as pd

dataset_size=10
N=32
M=64 # sample number
batch_size=1 # batch data among total data (dataset_size)
width_size=64
depth=2
seed=5678

key = jax.random.PRNGKey(seed)
data_key, model_key, loader_key = jax.random.split(key, 3)

# data
length = 50
ts = jnp.linspace(0.0, 10.0, length)
ys = jnp.ones((10,length,1)) # batch, length, data size
_, length_size, data_size = ys.shape

class Func(eqx.Module):
    mlp: eqx.nn.MLP

    def __init__(self, data_size, width_size, depth, *, key, **kwargs):
        super().__init__(**kwargs)
        self.mlp = eqx.nn.MLP(
            in_size = data_size,
            out_size = data_size,
            width_size = width_size,
            depth = depth,
            activation = jnp.tanh,
            use_bias = True, # remove bias (False)
            use_final_bias = True,
            key = key,
        )

    def __call__(self, t, y, args):
        return self.mlp(y)
    
model = Func(data_size, width_size, depth, key=model_key) # MLP model

def probability(b):
    p = b ** 2
    return p

# L : Time length, D : Data dimension, batch_num : number of batch
@eqx.filter_jit
def calc_sample(ti, yi, key):
    # ti : (L,)
    # yi : (L,D)

    b = jax.vmap(model.mlp)(yi) # (L,D)
    approx_p = jax.vmap( probability )(b) # (L,D)

    return approx_p

@eqx.filter_jit
def calc_prob(ti, yi):
    # calculate sample mean probability for each batch
    # ti : (L,)
    # yi : (L,D)

    start = time.time()
    subkeys = jax.random.split(key, M) 
    sample_p = jax.vmap(calc_sample, in_axes=(None,None,0))(ti,yi,subkeys) # (M,L,D,1) 
    # sample_p = calc_sample(ti,yi,key) 
    end = time.time()
    
    print("Time : ", end-start)
    sample_p_mean = jnp.sum(sample_p)

    return sample_p_mean

# In this case, I get weights of layers are all zero.
@eqx.filter_value_and_grad
def grad_loss(model, ti, yi):
    # yi : (batch_num, L=length * length_size, data size)
    # ti : (L,)
    # approximate the probability

    # Sample M
    p_pred = jax.vmap(calc_prob, in_axes=(None,0))(ti,yi) # along batch_num
    return jnp.sum( p_pred ) # negative log likelihood

@eqx.filter_jit
def make_step(ti, yi, model, opt_state):
    loss, grads = grad_loss(model, ti, yi)
    updates, opt_state = optim.update(grads, opt_state)
    model = eqx.apply_updates(model, updates)
    return loss, model, opt_state, grads

lr = 1e-3
steps = 100
optim = optax.adabelief(lr)
opt_state = optim.init(eqx.filter(model, eqx.is_inexact_array))

for step in range(steps):
    loss, model, opt_state, grads = make_step(ts, ys, model, opt_state)
    a,b,c = grads.mlp.layers
    print(a.weight[:3]) # check the gradients of weights in layers
```

![image](https://github.com/patrick-kidger/equinox/assets/54544508/a287e52b-3a39-439d-bb87-31e12c27049b)

It only depends on how to calculate the loss. In first one, I just calculate the loss along the batch size and sample size (because I want to calculate many sample losses) but I break it down into small functions such as calc_prob and calc_sample.

Thank you for your help.

**Comment by patrick-kidger** [comment]
I see the issue. In this function:
```python
@eqx.filter_value_and_grad
def grad_loss(model, ti, yi):
    # yi : (batch_num, L=length * length_size, data size)
    # ti : (L,)
    # approximate the probability

    # Sample M
    p_pred = jax.vmap(calc_prob, in_axes=(None,0))(ti,yi) # along batch_num
    return jnp.sum( p_pred ) # negative log likelihood
```
you're not actually using the `model` argument!

Thinking mathematically: the function `grad_loss` doesn't change as `model` varies. So by definition its derivative is zero.

**Comment by Kyung-mi** [comment]
Thank you for response. I should have added argument for model.
Thank you so much!
